### PR TITLE
upstream: Honor retry budget defaults for non-runtime configs

### PR DESCRIPTION
**Pull request by tonya11en**
Until now, retry budget configurations only setting default values if
the parameters were partially configured via runtime. For non-runtime
configs, such as xDS updates with a partially set retry budget field,
the default values were not being set. This patch forces the upstream
implementation to set default values for all methods of configuration.

Risk level:
Low

Testing:
Existing unit tests.

Fixes: #14842

### CODE REVIEW

File: source/common/upstream/upstream_impl.cc [hunk 1]
```diff
@@ -1241,17 +1241,13 @@ ClusterInfoImpl::ResourceManagers::load(const envoy::config::cluster::v3::Cluste
...
-      budget_percent = it->retry_budget().has_budget_percent()
-                           ? PROTOBUF_GET_WRAPPED_REQUIRED(it->retry_budget(), budget_percent)
-                           : budget_percent;
-      min_retry_concurrency =
-          it->retry_budget().has_min_retry_concurrency()
-              ? PROTOBUF_GET_WRAPPED_REQUIRED(it->retry_budget(), min_retry_concurrency)
-              : min_retry_concurrency;
+      // The budget_percent and min_retry_concurrency values are only set if there is a retry budget
+      // message set in the cluster config.
+      budget_percent = PROTOBUF_GET_WRAPPED_OR_DEFAULT(
```
Comment (tonya11en) [inline source/common/upstream/upstream_impl.cc:1246]: I looked around trying to find a way to test this, but the existing upstream and resource manager UTs didn't seem to have a way to test `ClusterInfoImpl::ResourceManagers::load()`. I'm a bit rusty in this area, so if there's a pointer to a test that does something similar I'd be glad to implement one to specifically test for this.

File: source/common/upstream/resource_manager_impl.h [hunk 2]
```diff
@@ -99,6 +99,9 @@ class ResourceManagerImpl : public ResourceManager {
   ResourceLimit& retries() override { return retries_; }
   ResourceLimit& connectionPools() override { return connection_pools_; }
 
+  static double defaultBudgetPercent() { return 20.0; }
+  static uint32_t defaultMinRetryConcurrency() { return 3; }
```
Comment (snowp) [inline source/common/upstream/resource_manager_impl.h:103]: Maybe just make these static values since they are just trivial types?

**Review by snowp — CHANGES_REQUESTED** [review]
Thanks this looks right, definitely want some kind of test coverage though.

@mpuncel does this seem reasonable to you?

**Comment by mpuncel** [comment]
This looks great! thanks for the fix!

File: test/common/upstream/upstream_impl_test.cc [hunk 3]
```diff
@@ -2276,6 +2284,71 @@ TEST_F(ClusterInfoImplTest, Metadata) {
...
+      - priority: DEFAULT
+        max_retries: 123
+      # 1 - Empty budget config, should have defaults.
+      - priority: DEFAULT
+        max_retries: 123
+        retry_budget: {}
+      # 2 - Empty budget config with max retries, should have budget defaults.
+      - priority: DEFAULT
+        max_retries: 123
+        retry_budget: {}
```
Comment (snowp) [inline test/common/upstream/upstream_impl_test.cc:2309]: These two seem identical?

**Review by snowp — CHANGES_REQUESTED** [review]
LGTM besides the one comment

**Comment by tonya11en** [comment]
Sorry, I had no idea the ball was in my court. Took care of it and I'll check back in a couple of hours to ensure the tests are passing.

**Review by snowp — APPROVED** [review]
Thanks!

/retest

**Pull request merged**